### PR TITLE
Fix compiler crash on xml attribute access on non `xml:Element` values

### LINKED ISSUE #35191: Using `xml-attribute-access-expr` with built-in subtypes of `xml` crashes at runtime
**Description:**
$title

**Steps to reproduce:**
```ballerina
public function main() {
    xml:Comment x5 = xml `<!--This is a comment-->`;
    io:println(x5.attr);

    // xml:ProcessingInstruction x5 = xml `<?xml-stylesheet type="text/xsl" href="style.xsl"?>`;
    // io:println(x5.attr);

    // xml:Text x5 = xml `This is a text`;
    // io:println(x5.attr);
}

```
Error log:
```
SEVERE {b7a.log.crash} - Bad type on operand stack
Exception Details:
  Location:
    main.main(Lio/ballerina/runtime/internal/scheduling/Strand;)Ljava/lang/Object; @405: invokevirtual
  Reason:
    Type 'io/ballerina/runtime/api/values/BString' (current frame, stack[1]) is not assignable to 'java/lang/String'
  Current Frame:
    bci: @405
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', 'io/ballerina/runtime/internal/values/XmlValue', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/internal/values/ArrayValue', long, long_2nd, null, null, null, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'java/lang/Object', 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/Object', integer }
    stack: { 'io/ballerina/runtime/internal/values/XmlValue', 'io/ballerina/runtime/api/values/BString' }
  Bytecode:
    0000000: 2ab4 0010 9900 07b8 0016 bf01 4c01 4d01
    0000010: 4e09 3704 013a 0601 3a07 013a 0801 3a09
    0000020: 0937 0a09 370c 013a 0e09 370f 0937 1101
    0000030: 3a13 0937 1409 3716 013a 1809 3719 0937
    0000040: 1b01 3a1d 013a 1e01 3a1f 0336 202a b400
    0000050: 1a9d 02eb 1520 ab00 0000 036a 0000 000c
    0000060: 0000 0000 0000 006a 0000 0001 0000 007e
    0000070: 0000 0002 0000 00e3 0000 0003 0000 016b
    0000080: 0000 0004 0000 0191 0000 0005 0000 01db
    0000090: 0000 0006 0000 01e3 0000 0007 0000 01fa
    00000a0: 0000 0008 0000 0241 0000 0009 0000 027f
    00000b0: 0000 000a 0000 029b 0000 000b 0000 02db
    00000c0: 1100 0036 20b2 0020 3a18 1400 2137 1914
    00000d0: 0023 371b 1100 0136 202a b600 2899 001d
    00000e0: 2a03 b500 2b2a b400 2fc6 000e 2ab4 002f
    00000f0: 592a 01b5 002f bfa7 0029 bb00 3159 2ab2
    0000100: 0037 b700 3ab2 003d 1918 1619 161b b200
    0000110: 41b2 0044 b200 47b2 004a b200 4ab8 0050
    0000120: 2ab6 0053 9a02 9ca7 0003 b200 203a 0914
    0000130: 0054 370a 1400 5637 0c11 0002 3620 2ab6
    0000140: 0028 9900 1d2a 03b5 002b 2ab4 002f c600
    0000150: 0e2a b400 2f59 2a01 b500 2fbf a700 1abb
    0000160: 0031 592a b200 37b7 003a b200 3d19 0916
    0000170: 0a16 0cb8 005b 2ab6 0053 9a02 46a7 0003
    0000180: b200 5e4d 2c03 b800 644c 1400 6537 04b2
    0000190: 0069 4d2b 2cb6 006f 3a07 1907 3a06 bb00
    00001a0: 7159 b200 7716 0414 0023 88bd 0079 5914
    00001b0: 007a 88bb 0079 5919 06b7 007e 53b7 0081
    00001c0: 4e11 0003 3620 2a2d 04b8 0087 3a08 2ab6
    00001d0: 0053 9a01 eea7 0003 b200 203a 0e14 0088
    00001e0: 370f 1400 2337 1111 0004 3620 2ab6 0028
    00001f0: 9900 1d2a 03b5 002b 2ab4 002f c600 0e2a
    0000200: b400 2f59 2a01 b500 2fbf a700 1abb 0031
    0000210: 592a b200 37b7 003a b200 3d19 0e16 0f16
    0000220: 11b8 005b 2ab6 0053 9a01 98a7 0003 013a
    0000230: 1f11 0005 3620 a700 0b11 0006 3620 a700
    0000240: 03b2 0020 3a13 1400 8837 1414 0023 3716
    0000250: 1100 0736 202a b600 2899 001d 2a03 b500
    0000260: 2b2a b400 2fc6 000e 2ab4 002f 592a 01b5
    0000270: 002f bfa7 001a bb00 3159 2ab2 0037 b700
    0000280: 3ab2 003d 1913 1614 1616 b800 5b2a b600
    0000290: 539a 012f a700 0311 0008 3620 2ab6 0028
    00002a0: 9900 1d2a 03b5 002b 2ab4 002f c600 0e2a
    00002b0: b400 2f59 2a01 b500 2fbf a700 11bb 0031
    00002c0: 592a b200 37b7 003a b800 8d2a b600 539a
    00002d0: 00f1 a700 0311 0009 3620 191f b000 00bf
    00002e0: 3a1d a700 08b8 0095 3a1d 191d c000 8f3a
    00002f0: 1e11 000a 3620 2ab6 0028 9900 1d2a 03b5
    0000300: 002b 2ab4 002f c600 0e2a b400 2f59 2a01
    0000310: b500 2fbf a700 13bb 0031 592a b200 37b7
    0000320: 003a 191e b800 992a b600 539a 0095 a700
    0000330: 0311 000b 3620 191d c000 9bbf 2ab4 009f
    0000340: 2a59 b400 1a04 645a b500 1a32 c000 a159
    0000350: b400 a43a 1f59 b400 a84c 59b4 00ab 4d59
    0000360: b400 af4e 59b4 00b2 3a09 59b4 00b6 370a
    0000370: 59b4 00b9 370c 59b4 00bc 3a0e 59b4 00bf
    0000380: 370f 59b4 00c2 3711 59b4 00c5 3a13 59b4
    0000390: 00c8 3714 59b4 00cb 3716 59b4 00ce 3a18
    00003a0: 59b4 00d1 3719 59b4 00d4 371b 59b4 00d7
    00003b0: 3a1d 59b4 00db 3a1e b400 de36 20a7 fc97
    00003c0: bb00 a159 b700 df59 191f b500 a459 2bb5
    00003d0: 00a8 592c b500 ab59 2db5 00af 5919 09b5
    00003e0: 00b2 5916 0ab5 00b6 5916 0cb5 00b9 5919
    00003f0: 0eb5 00bc 5916 0fb5 00bf 5916 11b5 00c2
    0000400: 5919 13b5 00c5 5916 14b5 00c8 5916 16b5
    0000410: 00cb 5919 18b5 00ce 5916 19b5 00d1 5916
    0000420: 1bb5 00d4 5919 1db5 00d7 5919 1eb5 00db
    0000430: 5915 20b5 00de 3a26 2ab4 009f 2a59 b400
    0000440: 1a5a 0460 b500 1a19 2653 191f b0       
  Exception Handler Table:
    bci [192, 733] => handler: 736
    bci [192, 733] => handler: 741
  Stackmap Table:
    same_frame(@11)
    full_frame(@84,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Null,Null,Null,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame_extended(@192)
    same_frame(@212)
    same_frame(@247)
    same_frame(@250)
    same_frame(@288)
    same_frame(@298)
    same_frame(@313)
    same_frame(@348)
    same_frame(@351)
    same_frame(@374)
    same_frame(@384)
    full_frame(@449,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Null,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    full_frame(@472,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame(@487)
    same_frame(@522)
    same_frame(@525)
    same_frame(@548)
    same_frame(@558)
    same_frame(@561)
    full_frame(@569,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Null,Null,Null,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    full_frame(@577,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame(@592)
    same_frame(@627)
    same_frame(@630)
    same_frame(@653)
    same_frame(@663)
    same_frame(@698)
    same_frame(@701)
    same_frame(@715)
    same_frame(@725)
    full_frame(@733,{},{Object[#232]})
    full_frame(@736,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{Object[#143]})
    same_locals_1_stack_item_frame(@741,Object[#145])
    full_frame(@746,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#143],Object[#143],Object[#4],Integer},{})
    full_frame(@753,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame(@788)
    same_frame(@791)
    same_frame(@807)
    same_frame(@817)
    full_frame(@828,{Object[#12],Null,Null,Null,Long,Null,Null,Null,Null,Long,Long,Null,Long,Long,Null,Long,Long,Null,Long,Long,Null,Null,Null,Integer},{})
    full_frame(@960,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
 
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    main.main(Lio/ballerina/runtime/internal/scheduling/Strand;)Ljava/lang/Object; @405: invokevirtual
  Reason:
    Type 'io/ballerina/runtime/api/values/BString' (current frame, stack[1]) is not assignable to 'java/lang/String'
  Current Frame:
    bci: @405
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', 'io/ballerina/runtime/internal/values/XmlValue', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/internal/values/ArrayValue', long, long_2nd, null, null, null, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'io/ballerina/runtime/api/values/BString', long, long_2nd, long, long_2nd, 'java/lang/Object', 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/Object', integer }
    stack: { 'io/ballerina/runtime/internal/values/XmlValue', 'io/ballerina/runtime/api/values/BString' }
  Bytecode:
    0000000: 2ab4 0010 9900 07b8 0016 bf01 4c01 4d01
    0000010: 4e09 3704 013a 0601 3a07 013a 0801 3a09
    0000020: 0937 0a09 370c 013a 0e09 370f 0937 1101
    0000030: 3a13 0937 1409 3716 013a 1809 3719 0937
    0000040: 1b01 3a1d 013a 1e01 3a1f 0336 202a b400
    0000050: 1a9d 02eb 1520 ab00 0000 036a 0000 000c
    0000060: 0000 0000 0000 006a 0000 0001 0000 007e
    0000070: 0000 0002 0000 00e3 0000 0003 0000 016b
    0000080: 0000 0004 0000 0191 0000 0005 0000 01db
    0000090: 0000 0006 0000 01e3 0000 0007 0000 01fa
    00000a0: 0000 0008 0000 0241 0000 0009 0000 027f
    00000b0: 0000 000a 0000 029b 0000 000b 0000 02db
    00000c0: 1100 0036 20b2 0020 3a18 1400 2137 1914
    00000d0: 0023 371b 1100 0136 202a b600 2899 001d
    00000e0: 2a03 b500 2b2a b400 2fc6 000e 2ab4 002f
    00000f0: 592a 01b5 002f bfa7 0029 bb00 3159 2ab2
    0000100: 0037 b700 3ab2 003d 1918 1619 161b b200
    0000110: 41b2 0044 b200 47b2 004a b200 4ab8 0050
    0000120: 2ab6 0053 9a02 9ca7 0003 b200 203a 0914
    0000130: 0054 370a 1400 5637 0c11 0002 3620 2ab6
    0000140: 0028 9900 1d2a 03b5 002b 2ab4 002f c600
    0000150: 0e2a b400 2f59 2a01 b500 2fbf a700 1abb
    0000160: 0031 592a b200 37b7 003a b200 3d19 0916
    0000170: 0a16 0cb8 005b 2ab6 0053 9a02 46a7 0003
    0000180: b200 5e4d 2c03 b800 644c 1400 6537 04b2
    0000190: 0069 4d2b 2cb6 006f 3a07 1907 3a06 bb00
    00001a0: 7159 b200 7716 0414 0023 88bd 0079 5914
    00001b0: 007a 88bb 0079 5919 06b7 007e 53b7 0081
    00001c0: 4e11 0003 3620 2a2d 04b8 0087 3a08 2ab6
    00001d0: 0053 9a01 eea7 0003 b200 203a 0e14 0088
    00001e0: 370f 1400 2337 1111 0004 3620 2ab6 0028
    00001f0: 9900 1d2a 03b5 002b 2ab4 002f c600 0e2a
    0000200: b400 2f59 2a01 b500 2fbf a700 1abb 0031
    0000210: 592a b200 37b7 003a b200 3d19 0e16 0f16
    0000220: 11b8 005b 2ab6 0053 9a01 98a7 0003 013a
    0000230: 1f11 0005 3620 a700 0b11 0006 3620 a700
    0000240: 03b2 0020 3a13 1400 8837 1414 0023 3716
    0000250: 1100 0736 202a b600 2899 001d 2a03 b500
    0000260: 2b2a b400 2fc6 000e 2ab4 002f 592a 01b5
    0000270: 002f bfa7 001a bb00 3159 2ab2 0037 b700
    0000280: 3ab2 003d 1913 1614 1616 b800 5b2a b600
    0000290: 539a 012f a700 0311 0008 3620 2ab6 0028
    00002a0: 9900 1d2a 03b5 002b 2ab4 002f c600 0e2a
    00002b0: b400 2f59 2a01 b500 2fbf a700 11bb 0031
    00002c0: 592a b200 37b7 003a b800 8d2a b600 539a
    00002d0: 00f1 a700 0311 0009 3620 191f b000 00bf
    00002e0: 3a1d a700 08b8 0095 3a1d 191d c000 8f3a
    00002f0: 1e11 000a 3620 2ab6 0028 9900 1d2a 03b5
    0000300: 002b 2ab4 002f c600 0e2a b400 2f59 2a01
    0000310: b500 2fbf a700 13bb 0031 592a b200 37b7
    0000320: 003a 191e b800 992a b600 539a 0095 a700
    0000330: 0311 000b 3620 191d c000 9bbf 2ab4 009f
    0000340: 2a59 b400 1a04 645a b500 1a32 c000 a159
    0000350: b400 a43a 1f59 b400 a84c 59b4 00ab 4d59
    0000360: b400 af4e 59b4 00b2 3a09 59b4 00b6 370a
    0000370: 59b4 00b9 370c 59b4 00bc 3a0e 59b4 00bf
    0000380: 370f 59b4 00c2 3711 59b4 00c5 3a13 59b4
    0000390: 00c8 3714 59b4 00cb 3716 59b4 00ce 3a18
    00003a0: 59b4 00d1 3719 59b4 00d4 371b 59b4 00d7
    00003b0: 3a1d 59b4 00db 3a1e b400 de36 20a7 fc97
    00003c0: bb00 a159 b700 df59 191f b500 a459 2bb5
    00003d0: 00a8 592c b500 ab59 2db5 00af 5919 09b5
    00003e0: 00b2 5916 0ab5 00b6 5916 0cb5 00b9 5919
    00003f0: 0eb5 00bc 5916 0fb5 00bf 5916 11b5 00c2
    0000400: 5919 13b5 00c5 5916 14b5 00c8 5916 16b5
    0000410: 00cb 5919 18b5 00ce 5916 19b5 00d1 5916
    0000420: 1bb5 00d4 5919 1db5 00d7 5919 1eb5 00db
    0000430: 5915 20b5 00de 3a26 2ab4 009f 2a59 b400
    0000440: 1a5a 0460 b500 1a19 2653 191f b0       
  Exception Handler Table:
    bci [192, 733] => handler: 736
    bci [192, 733] => handler: 741
  Stackmap Table:
    same_frame(@11)
    full_frame(@84,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Null,Null,Null,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame_extended(@192)
    same_frame(@212)
    same_frame(@247)
    same_frame(@250)
    same_frame(@288)
    same_frame(@298)
    same_frame(@313)
    same_frame(@348)
    same_frame(@351)
    same_frame(@374)
    same_frame(@384)
    full_frame(@449,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Null,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    full_frame(@472,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame(@487)
    same_frame(@522)
    same_frame(@525)
    same_frame(@548)
    same_frame(@558)
    same_frame(@561)
    full_frame(@569,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Null,Null,Null,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    full_frame(@577,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame(@592)
    same_frame(@627)
    same_frame(@630)
    same_frame(@653)
    same_frame(@663)
    same_frame(@698)
    same_frame(@701)
    same_frame(@715)
    same_frame(@725)
    full_frame(@733,{},{Object[#232]})
    full_frame(@736,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{Object[#143]})
    same_locals_1_stack_item_frame(@741,Object[#145])
    full_frame(@746,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#143],Object[#143],Object[#4],Integer},{})
    full_frame(@753,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})
    same_frame(@788)
    same_frame(@791)
    same_frame(@807)
    same_frame(@817)
    full_frame(@828,{Object[#12],Null,Null,Null,Long,Null,Null,Null,Null,Long,Long,Null,Long,Long,Null,Long,Long,Null,Long,Long,Null,Null,Null,Integer},{})
    full_frame(@960,{Object[#12],Object[#107],Object[#228],Object[#230],Long,Object[#107],Object[#107],Object[#4],Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#228],Long,Long,Object[#4],Object[#143],Object[#4],Integer},{})

        at $_init.$lambda$main$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:611)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:317)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:284)
        at java.base/java.lang.Thread.run(Thread.java:834)
```
**Affected Versions:**
Ballerina 2201.0.0 (Swan Lake)

### PR DESCRIPTION
**Pull request by poorna2152**
Fixes #35191 

## Approach
Fix xml attribute access on all subtypes of xml

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -8619,6 +8619,9 @@ private BType getLaxFieldAccessType(BType exprType) {
                 return symTable.jsonType;
             case TypeTags.XML:
             case TypeTags.XML_ELEMENT:
+            case TypeTags.XML_COMMENT:
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:8622]: My concern is why we include XML within the `getLaxFieldAccessType` function. If the type is lax, then the lax field access rules are included. XML is not a lax type, correct?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -314,7 +314,7 @@ public BType checkType(Location pos,
     public boolean isLaxFieldAccessAllowed(BType type) {
         type = Types.getImpliedType(type);
         Set<BType> visited = new HashSet<>();
-        return isLaxType(type, visited) == 1 || type.tag == TypeTags.XML || type.tag == TypeTags.XML_ELEMENT;
+        return isLaxType(type, visited) == 1 || isAssignable(type, symTable.xmlType);
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:317]: In the previous implementation, this was also incorrect. Lax field access should only be allowed for lax types. However, XML is not a lax type. I have concerns about including this logic within this function.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 1 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:8622]: As per the offline discussion, even though xml is not considered as lax type, lax typing rules apply for xml attribute access. I guess that is the reason for having xml type here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 2 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:317]: Similar reason to the above comment, as per the offline discussion

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAttributeAccessTest.java [hunk 3]
```diff
@@ -114,6 +114,16 @@ public void testErrorsOnXMLIndexedOptionalAttributeAccess() {
...
+    @Test
+    public void testXmlAttributeAccessOnXmlUnionTypes() {
+        BRunUtil.invoke(compileResult, "testXmlAttributeAccessOnXmlUnionTypes");
+    }
+
+    @Test
+    public void testErrorsOnXmlAttributeAccessOnNonXmlElementValue() {
+        BRunUtil.invoke(compileResult, "testErrorsOnXmlAttributeAccessOnNonXmlElementValue");
+    }
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAttributeAccessTest.java:126]: Please add some negative tests too.

E.g.,

```ballerina
xml:Comment x = xml `<!--This is a comment-->`;
xml xmlResult = x.attr; // error
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -6377,7 +6377,7 @@ private void rewriteFieldBasedAccess(BLangFieldBasedAccess fieldAccessExpr) {
                         (BVarSymbol) fieldAccessExpr.symbol, false, fieldAccessExpr.isStoreOnCreation);
             }
         } else if (types.isLaxFieldAccessAllowed(refType)) {
-            if (!(varRefTypeTag == TypeTags.XML || varRefTypeTag == TypeTags.XML_ELEMENT)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6380]: There seems to be an issue in the type set for optional attribute access.

```ballerina
xml:Comment x = xml `<!--This is a comment-->`;
string? xmlResult = x?.attr; // should result in a compile-time error
io:println(<any|error> xmlResult is error); // true, although static type doesn't allow error
```

Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6380]: Should this line result in a compile time error?
```ballerina
string? xmlResult = x?.attr; // should result in a compile-time error
```
I'm assuming the reason this should be a compile time error is that the static type of the right hand side should be `string|error?`.
But as per the spec, https://ballerina.io/spec/lang/master/#section_6.12 the optional attribute access results in an error when `if v is not a singleton element, the result is an error. `  if the element x here is not a singleton. But in this case we know that the element is a singleton. So should this still be a compile time error?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6380]: That explanation is for runtime behaviour. The static type is defined simply as

> The static type of an xml-required-attribute-access-expr is string|error. The static type of an xml-optional-attribute-access-expr is string|error|(). 

So the implementation has to conform to this.

Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6380]: Ack. Thanks
This seems to exists the existing implementation as well.
```ballerina
function foo() {
    xml:Element x = xml `<x/>`;
    string? xmlResult = x?.attr;
}
```
Will update the fix

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-attribute-access-syntax.bal [hunk 5]
```diff
@@ -194,6 +194,146 @@ function testErrorsOnXMLIndexedOptionalAttributeAccess() {
...
 
+type XC xml:Comment;
+type XPI xml:ProcessingInstruction;
+type XT xml:Text;
+type XE xml:Element;
+
+function testXmlAttributeAccessOnXmlUnionTypes() {
+    xml<xml:Element|xml:Text> x1 = xml `<a attr="aa">a</a>`;
+    assertNonErrorValue(x1.attr, "aa");
+    assertNonErrorValue(x1?.attr, "aa");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-attribute-access-syntax.bal:205]: At least for a few of these, we have to declare variables and test. Or we won't be testing typing positive cases.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-attribute-access-syntax.bal:205]: Same for testErrorsOnXmlAttributeAccessOnNonXmlElementValue.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**